ext/openssl/ossl_ts.c

```diff
@@ -0,0 +1,1492 @@
+ * must be set.
+ *
+ * Raises a TimestampError if creation fails.
+ *
+ * call-seq:
+ *       factory.create_timestamp(key, certificate, request) -> Response
+ */
+static VALUE
+ossl_tsfac_create_ts(VALUE self, VALUE key, VALUE certificate, VALUE request)
+{
+    VALUE serial_number, def_policy_id, gen_time, additional_certs;
+    VALUE str, cert;
+    STACK_OF(X509) *inter_certs;
+    VALUE ret;
+    EVP_PKEY *sign_key;
+    X509 *tsa_cert;
+    TS_REQ *req;
+    TS_RESP *response = NULL;
+    TS_RESP_CTX *ctx = NULL;
+    BIO *req_bio;
+    ASN1_INTEGER *asn1_serial = NULL;
+    ASN1_OBJECT *def_policy_id_obj = NULL;
+    long lgen_time;
+    const char * err_msg = NULL;
+    int i, status = 0;
+
+    tsa_cert = GetX509CertPtr(certificate);
+    sign_key = GetPrivPKeyPtr(key);
+    GetTSRequest(request, req);
+
+    gen_time = ossl_tsfac_get_gen_time(self);
+    if (!rb_obj_is_instance_of(gen_time, rb_cTime)) {
+	err_msg = "@gen_time must be a Time.";
+	goto end;
+    }
+    lgen_time = NUM2LONG(rb_funcall(gen_time, rb_intern("to_i"), 0));
+
+    serial_number = ossl_tsfac_get_serial_number(self);
+    if (serial_number == Qnil) {
+	err_msg = "@serial_number must be set.";
+	goto end;
+    }
+    asn1_serial = num_to_asn1integer(serial_number, NULL);
+
+    def_policy_id = ossl_tsfac_get_default_policy_id(self);
+    if (def_policy_id == Qnil && !TS_REQ_get_policy_id(req)) {
+	err_msg = "No policy id in the request and no default policy set";
+	goto end;
+    }
+    if (def_policy_id != Qnil && !TS_REQ_get_policy_id(req))
+	def_policy_id_obj = obj_to_asn1obj(def_policy_id);
+
+    if (!(ctx = TS_RESP_CTX_new())) {
+	err_msg = "Memory allocation failed.";
+	goto end;
+    }
+
+    TS_RESP_CTX_set_serial_cb(ctx, ossl_tsfac_serial_cb, asn1_serial);
+    if (!TS_RESP_CTX_set_signer_cert(ctx, tsa_cert)) {
+	err_msg = "Certificate does not contain the timestamping extension";
+	goto end;
+    }
+
+    additional_certs = ossl_tsfac_get_additional_certs(self);
+    if (rb_obj_is_kind_of(additional_certs, rb_cArray)) {
+	inter_certs = ossl_protect_x509_ary2sk(additional_certs, &status);
+	if (status)
+		goto end;
+
+	/* this dups the sk_X509 and ups each cert's ref count */
+	TS_RESP_CTX_set_certs(ctx, inter_certs);
+	sk_X509_pop_free(inter_certs, X509_free);
+    }
+
+    TS_RESP_CTX_set_signer_key(ctx, sign_key);
+    if (def_policy_id != Qnil && !TS_REQ_get_policy_id(req))
+	TS_RESP_CTX_set_def_policy(ctx, def_policy_id_obj);
+    if (TS_REQ_get_policy_id(req))
+	TS_RESP_CTX_set_def_policy(ctx, TS_REQ_get_policy_id(req));
+    TS_RESP_CTX_set_time_cb(ctx, ossl_tsfac_time_cb, &lgen_time);
+
+    TS_RESP_CTX_add_md(ctx, EVP_md5());
+    TS_RESP_CTX_add_md(ctx, EVP_sha1());
+    TS_RESP_CTX_add_md(ctx, EVP_sha224());
+    TS_RESP_CTX_add_md(ctx, EVP_sha256());
+    TS_RESP_CTX_add_md(ctx, EVP_sha384());
+    TS_RESP_CTX_add_md(ctx, EVP_sha512());
+
+    str = rb_funcall(request, rb_intern("to_der"), 0);
+    req_bio = ossl_obj2bio(&str);
```

Comment:
rb_funcall() and ossl_obj2bio() can raise an exception. By the way, the rb_funcall() can be replaced by `ossl_to_der(request)` (this still raises an exception).